src/wallet/scriptpubkeyman.cpp

```diff
@@ -0,0 +1,597 @@
+{
+    LOCK(wallet->cs_wallet);
+    assert(m_max_keypool_index < std::numeric_limits<int64_t>::max()); // How in the hell did you use so many keys?
+    int64_t index = ++m_max_keypool_index;
+    if (!batch.WritePool(index, CKeyPool(pubkey, type))) {
+        throw std::runtime_error(std::string(__func__) + ": writing imported pubkey failed");
+    }
+
+    switch (type) {
+        case HDChain::ChangeType::EXTERNAL:
+            setExternalKeyPool.insert(index);
+            break;
+        case HDChain::ChangeType::INTERNAL:
+            setInternalKeyPool.insert(index);
+            break;
+        case HDChain::ChangeType::STAKING:
+            setStakingKeyPool.insert(index);
+            break;
+    }
+
+    m_pool_key_to_index[pubkey.GetID()] = index;
+}
+
+/**
+ * Generate a new key and stores it in db.
+ */
+CPubKey ScriptPubKeyMan::GenerateNewKey(CWalletDB &batch, const uint8_t& type)
+{
+    AssertLockHeld(wallet->cs_wallet);
+    bool fCompressed = wallet->CanSupportFeature(FEATURE_COMPRPUBKEY); // default to compressed public keys if we want 0.6.0 wallets
+
+    CKey secret;
+
+    // Create new metadata
+    int64_t nCreationTime = GetTime();
+    CKeyMetadata metadata(nCreationTime);
+
+    // use HD key derivation if HD was enabled during wallet creation and a seed is present
+    if (IsHDEnabled()) {
+        DeriveNewChildKey(batch, metadata, secret, type);
+    } else {
+        secret.MakeNewKey(fCompressed);
+    }
+
+    // Compressed public keys were introduced in version 0.6.0
+    if (fCompressed) {
+        wallet->SetMinVersion(FEATURE_COMPRPUBKEY);
+    }
+
+    CPubKey pubkey = secret.GetPubKey();
+    assert(secret.VerifyPubKey(pubkey));
+
+    wallet->mapKeyMetadata[pubkey.GetID()] = metadata;
+    UpdateTimeFirstKey(nCreationTime);
+
+    if (!AddKeyPubKeyWithDB(batch, secret, pubkey)) {
+        throw std::runtime_error(std::string(__func__) + ": AddKey failed");
+    }
+    return pubkey;
+}
+
+void ScriptPubKeyMan::DeriveNewChildKey(CWalletDB &batch, CKeyMetadata& metadata, CKey& secret, const uint8_t& changeType)
+{
+    AssertLockHeld(wallet->cs_wallet);
+    // Use BIP44 keypath scheme i.e. m / purpose' / coin_type' / account' / change / address_index
+    CKey seed;                     //seed (256bit)
+    CExtKey masterKey;             //hd master key
+    CExtKey purposeKey;            //key at m/purpose' --> key at m/44'
+    CExtKey cointypeKey;           //key at m/purpose'/coin_type'  --> key at m/44'/119'
+    CExtKey accountKey;            //key at m/purpose'/coin_type'/account' ---> key at m/44'/119'/account_num
+    CExtKey changeKey;             //key at m/purpose'/coin_type'/account'/change ---> key at m/44'/119'/account_num/change', external = 0' or internal = 1'.
+    CExtKey childKey;              //key at m/purpose'/coin_type'/account'/change/address_index ---> key at m/44'/119'/account_num/change'/<n>'
```

Comment:
nit: in the comments, account number is hardened as per BIP44 (`account_num'`)

Reply:
done